### PR TITLE
Nav 58 graph visualisation

### PR DESCRIPTION
**Pull request by jsoppela**

### CODE REVIEW

**Comment by jonathansberry** [comment]
This is a good start. :trophy:  Let's leave the PR open and come back to it when it floats back up to the top of the priority list.

**Comment by jonathansberry** [comment]
@jsoppela To get mypy to pass you need to remove dash from the type hinting checks. Look at this config file to see how I have done it for other dependancies: 

https://github.com/fjelltopp/navigator_engine/blob/development/mypy.ini 

If you have rebased onto the latest development branch and run `pipenv sync --dev` you should then be able to run `pipenv run mypy .` to check mypy runs locally - it'll be quicker than waiting for github actions each time.

File: navigator_engine/common/dash_app.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import dash
+import flask
+from navigator_engine.common import graph_visualizer
+
+
+def add_dash_app(navigator_app):
```
Comment (jonathansberry) [inline navigator_engine/common/dash_app.py:6]: Type hint?

File: navigator_engine/common/dash_app.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    dash_app = dash.Dash(__name__, server=navigator_app, url_base_pathname='/graph/')
+    graph_visualizer.create_visualizer(flask_app=navigator_app, dash_app=dash_app)
+
+    @navigator_app.route('/graph')
+    def graph():
+        return flask.redirect('/graph')
```
Comment (jonathansberry) [inline navigator_engine/common/dash_app.py:12]: I don't understand what is happening here... can you help me out?  It looks like you are putting redirect at this url to itself?  I would expect this to cause an eternal loop of redirects.

Where/how in this little endpoint does it hook into the dash app?

File: navigator_engine/common/graph_visualizer.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        html.Hr(),
+        html.Div(
+            id='cytoscape-container',
+            children=[
+                cyto.Cytoscape(
+                    id='cytoscape-figure',
+                    layout={'name': 'preset'},
+                    style={'width': '100%', 'height': '400px'},
+                    elements=[
+                    ])
```
Comment (jonathansberry) [inline navigator_engine/common/graph_visualizer.py:93]: I know it is picky, but I really do like all my brackets aligned with the same indentation.  Can you move the square bracket onto the line above and put the parenthesis in line with cyto.Cytoscape above?

File: navigator_engine/common/graph_visualizer.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            node = model.load_node(node_id=n.id)
+            if node.action and node.action.complete:
+                node_element = {'data': {'id': str(node.id),
+                                         'label': f'A {node.action.id} COMPLETE',
+                                         'infobox': f'Action {node.action.id} | '
+                                                    f'Node {node.id} | '
+                                                    f'{node.action.title}'
+                                         },
+                                'classes': 'red triangle'
+                                }
```
Comment (jonathansberry) [inline navigator_engine/common/graph_visualizer.py:151]: Can you format this dictionary and all dictionaries below like this please: 
```suggestion
                node_element = {
                    'classes': 'red triangle',
                    'data': {
                        'id': str(node.id),
                        'label': f'A {node.action.id} COMPLETE',
                        'infobox': f'Action {node.action.id} | '
                                   f'Node {node.id} | '
                                   f'{node.action.title}'
                    }
                 }
```

File: navigator_engine/common/graph_visualizer.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            else:
+                continue
+
+            elements.append(node_element)
+
+        for e in graph_x.edges:
+            edge = model.load_edge(from_id=e[0].id, to_id=e[1].id)
+            if edge.type:
+                edge_element = {'data': {'source': str(e[0].id), 'target': str(e[1].id)},
+                                'classes': 'true'}
```
Comment (jonathansberry) [inline navigator_engine/common/graph_visualizer.py:189]: I'm just not a fan of this "hanging" multiline format.  I don't think it's very readable.  The general principle is that the closing parenthesis should always be at the same indent as the line the block was opened on. The extra line creating a bit more white space is only a good thing in my head.  Doing it this way also shortens the line lengths which is good as well. 
```suggestion
                edge_element = {
                    'data': {'source': str(e[0].id), 'target': str(e[1].id)},
                    'classes': 'true'
                }
```

File: navigator_engine/common/graph_visualizer.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            elif not edge.type:
+                edge_element = {'data': {'source': str(e[0].id), 'target': str(e[1].id)},
+                                'classes': 'false'}
```
Comment (jonathansberry) [inline navigator_engine/common/graph_visualizer.py:192]: Same here.

File: navigator_engine/common/graph_visualizer.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+            else:
+                continue
+
+            elements.append(edge_element)
+
+        fig = cyto.Cytoscape(
+            id='cytoscape-figure',
+            layout={'name': 'breadthfirst',
+                    'roots': f'[id = "{root_nodes[0].id}"]'},
```
Comment (jonathansberry) [inline navigator_engine/common/graph_visualizer.py:201]: And here

File: navigator_engine/common/graph_visualizer.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+                                         'label': f'C {node.conditional.id}',
+                                         'infobox': f'Conditional '
+                                                    f'{node.conditional.id} | '
+                                                    f'Node {node.id}  | '
+                                                    f'{node.conditional.title}'
+                                         },
+                                'classes': 'green circle'
+                                }
+            else:
+                continue
```
Comment (jonathansberry) [inline navigator_engine/common/graph_visualizer.py:181]: Nice :ok_hand:

File: navigator_engine/common/graph_visualizer.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+def create_visualizer(flask_app, dash_app) -> None:
+
+    dash_app.layout = html.Div(children=[
+        html.Div(
+            id='dropdown-container',
+            children=[
+                dcc.Dropdown(
+                    id='graph-dropdown',
+                    options=[],
+                    placeholder='Select a graph',
```
Comment (jonathansberry) [inline navigator_engine/common/graph_visualizer.py:80]: Trailing comma?

File: navigator_engine/common/graph_visualizer.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
+import navigator_engine.model as model
+from dash import dcc
+import dash_cytoscape as cyto
+from dash import html
+from dash.dependencies import Input, Output
+
+
+graph_stylesheet = [  # Group selectors
```
Comment (jonathansberry) [inline navigator_engine/common/graph_visualizer.py:8]: Please can the comment just go on the line above to be consistent with the other comments?

**Review by jonathansberry — CHANGES_REQUESTED** [review]
Fantastic.  Thanks Jyri.  Just a few comments, mostly formatting related.

File: navigator_engine/common/dash_app.py (hunk 2 above)
Comment (jsoppela) [reply navigator_engine/common/dash_app.py]: The endpoint is configured in this line in dash_app.py: `dash_app = dash.Dash(__name__, server=navigator_app, url_base_pathname='/graph/')`

I removed the redirect you mentioned and it had no effect on the functionality. Seems like I just included without fully understanding the docs I was reading.

**Review by jonathansberry — APPROVED** [review]

**Pull request merged**